lte/gateway/python/magma/pipelined/app/middle.py

```diff
@@ -0,0 +1,189 @@
+            self._mtr_service_enabled = True
+            mtr_ip = config_dict['mtr_ip']
+            mtr_port = config_dict['ovs_mtr_port_number']
+            mtr_mac = get_virtual_iface_mac(config_dict['mtr_interface'])
+        else:
+            mtr_ip = None
+            mtr_mac = None
+            mtr_port = None
+
+        li_port_name = None
+        if 'li_local_iface' in config_dict:
+            li_port_name = config_dict['li_local_iface']
+
+        return self.MiddleConfig(
+            mtr_ip=mtr_ip,
+            mtr_port=mtr_port,
+            li_port_name=li_port_name,
+            setup_type=config_dict.get('setup_type', None),
+            mtr_mac=mtr_mac,
+        )
+
+    def _get_default_flow_msgs(self, datapath) -> DefaultMsgsMap:
+        return {
+            self._midle_tbl_num: self._get_default_middle_flow_msgs(datapath),
+        }
+
+    def _get_default_middle_flow_msgs(self, dp):
+        """
+        Egress table is the last table that a packet touches in the pipeline.
+        Output downlink traffic to gtp port, uplink traffic to LOCAL
+
+        Raises:
+            MagmaOFError if any of the default flows fail to install.
+        """
+        msgs = []
+        next_tbl = self._service_manager.get_next_table_num(PHYSICAL_TO_LOGICAL)
+
+        # Allow passthrough pkts(skip enforcement and send to egress table)
+        ps_match = MagmaMatch(passthrough=PASSTHROUGH_REG_VAL)
+        msgs.append(
+            flows.get_add_resubmit_next_service_flow_msg(
+                dp,
+                self._midle_tbl_num, ps_match, actions=[],
+                priority=flows.PASSTHROUGH_PRIORITY,
+                resubmit_table=self._egress_tbl_num,
+            ),
+        )
+
+        match = MagmaMatch()
+        msgs.append(
+            flows.get_add_resubmit_next_service_flow_msg(
+                dp,
+                self._midle_tbl_num, match, actions=[],
+                priority=flows.DEFAULT_PRIORITY, resubmit_table=next_tbl,
+            ),
+        )
+
+        if self._mtr_service_enabled:
+            msgs.extend(
+                get_vlan_egress_flow_msgs(
+                    dp,
+                    self._midle_tbl_num,
+                    ether_types.ETH_TYPE_IP,
+                    self.config.mtr_ip,
+                    self.config.mtr_port,
+                    priority=flows.UE_FLOW_PRIORITY,
+                    direction=Direction.OUT,
+                    dst_mac=self.config.mtr_mac,
+                ),
+            )
+        return msgs
+
+    def _install_default_flows(self, datapath):
+        default_msg_map = self._get_default_flow_msgs(datapath)
+        default_msgs = []
+
+        for _, msgs in default_msg_map.items():
+            default_msgs.extend(msgs)
+        chan = self._msg_hub.send(default_msgs, datapath)
+        self._wait_for_responses(chan, len(default_msgs))
+
+    def _wait_for_responses(self, chan, response_count):
```

Comment:
same as above, this function can be moved to base class.

Reply:
See answer above - additionally this function is not used by a lot of controllers that extend base.by.